docs/RFCS/20171003_interleaved_table_joins.md

```diff
@@ -0,0 +1,779 @@
+          * [Distributed execution engine (DistSQL)](#distributed-execution-engine-distsql)
+       * [Processors implementation](#processors-implementation)
+          * [Reading component](#reading-component)
+          * [Joining component](#joining-component)
+          * [Final output](#final-output)
+    * [Drawbacks](#drawbacks)
+    * [Rationale and Alternatives](#rationale-and-alternatives)
+       * [[1] Scanning the interleaved hierarchy](#1-scanning-the-interleaved-hierarchy)
+       * [[2] Logical planning &gt; physical planning](#2-logical-planning--physical-planning)
+       * [[3] Generalizing planning](#3-generalizing-planning)
+          * [[3a] Common ancestor joins](#3a-common-ancestor-joins)
+          * [[3b] Strict prefix joins](#3b-strict-prefix-joins)
+             * [Canonical prefix join example](#canonical-prefix-join-example)
+             * [Approach 1 (2-pass, one scan)](#approach-1-2-pass-one-scan)
+             * [Approach 2 (2-pass, least memory, O(2k) scans)](#approach-2-2-pass-least-memory-o2k-scans)
+          * [[3c] Subset prefix joins](#3c-subset-prefix-joins)
+       * [[4] Multi-table interleaved joins](#4-multi-table-interleaved-joins)
+       * [[5] Theta Joins](#5-theta-joins)
+    * [Unresolved questions](#unresolved-questions)
+
+# Summary
+
+We [currently permit
+users](https://www.cockroachlabs.com/docs/stable/interleave-in-parent.html) to
+specify if a child table should be interleaved on-disk under its specified
+parent table. This feature was adopted from [Google Spanner's interleaved
+tables](https://cloud.google.com/spanner/docs/schema-and-data-model#creating_interleaved_tables)
+and was meant to be an optimization the user could opt in for tables that are
+often queried in a parent-child relationship (i.e. one-to-one, one-to-many).
+Refer to the [RFC on interleaved
+tables](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20160624_sql_interleaved_tables.md)
+for more details.
+
+Beyond co-locating the child's key-value (kv) pairs with the parent's kv pairs
+in the underlying RocksDB storage, we do not currently take advantage of this
+fact to optimize queries.
+
+One type of query we can optimize for interleaved tables are joins between the
+parent and children tables. In the context of DistSQL, instead of spawning two
+`TableReader`s that scan twice, once for the parent and once for the child
+table, and a `Joiner` processor, we can do one scan over the relevant
+interleaved range as well as performing the join in tandem.
+
+The performance impact can be significant considering we reduce the number
+of gRPC streams between nodes and RocksDB scans.
+
+This RFC identifies two implementation phases for interleaved table joins:
+1. Planning interleaved table joins for a given query (Planning implementation phase)
+2. Mechanism that performs the joint scan as well as the join (Processors implementation phase)
+
+The Processors implementation phase will be carried out first.
```

Comment:
maybe order it first then?

Reply:
I had initially had it the other way around, but @knz brought up a good point that I should order it logically from top-down (planning --> processor).